### DISCUSSION TITLE
Getting error 167772427 wrong version number (SSL routines) c++ error in SSL_read

### DISCUSSION THREAD

**Discussion by dreamvrutik** [openssl/openssl#25846] [Q&A] [answered]
I am trying to do handshake using OpenSSL and memory bios in non blocking I/O. In order to be successful, I'm trying to mimic how boost asio internally works. I'm running a test where the server certificate chain length is 1 and client is 0. The handshake is succeeding but after that I get below error on both Server and client while doing SSL Read. I am attaching all the logs I printed:

[client] SSL where: 16 ret: 1 state: before SSL initialization
[client] SSL where: 4097 ret: 1 state: before SSL initialization
[client] SSL where: 4097 ret: 1 state: SSLv3/TLS write client hello
[client] SSL where: 4098 ret: -1 state: SSLv3/TLS write client hello
[b246df83a18436c2fd3ee572cf870586] PerformIO: Result: -1 sslError: 2 sysError: 0 pendingOutputBefore: 0 pendingOutputAfter: 517
[Client] [b246df83a18436c2fd3ee572cf870586] Performing SSL handshake: want_output
[Client] [b246df83a18436c2fd3ee572cf870586] Draining WriteBIO to socket
[server] SSL where: 16 ret: 1 state: before SSL initialization
[Client] [b246df83a18436c2fd3ee572cf870586] Drained WriteBIO to socket
[client] SSL where: 4098 ret: -1 state: SSLv3/TLS write client hello
[b246df83a18436c2fd3ee572cf870586] PerformIO: Result: -1 sslError: 2 sysError: 0 pendingOutputBefore: 0 pendingOutputAfter: 0
[server] SSL where: 8193 ret: 1 state: before SSL initialization
[Client] [b246df83a18436c2fd3ee572cf870586] Performing SSL handshake: want_input_and_retry
[Client] [b246df83a18436c2fd3ee572cf870586] Waiting for socket to be readable
[server] SSL where: 8194 ret: -1 state: before SSL initialization
[214419aac41463c6defd4348e9652eec] PerformIO: Result: -1 sslError: 2 sysError: 0 pendingOutputBefore: 0 pendingOutputAfter: 0
[Server] [214419aac41463c6defd4348e9652eec] Performing SSL handshake: want_input_and_retry
[Server] [214419aac41463c6defd4348e9652eec] Waiting for socket to be readable
[Server] [214419aac41463c6defd4348e9652eec] Read 517 bytes from socket
[Server] [214419aac41463c6defd4348e9652eec] Draining ReadBIO from unconsumedInput after socket read
[server] SSL where: 8193 ret: 1 state: before SSL initialization
[server] SSL where: 8193 ret: 1 state: SSLv3/TLS read client hello
[server] SSL where: 8193 ret: 1 state: SSLv3/TLS write server hello
[server] SSL where: 8193 ret: 1 state: SSLv3/TLS write change cipher spec
[server] SSL where: 8193 ret: 1 state: TLSv1.3 write encrypted extensions
[server] SSL where: 8193 ret: 1 state: SSLv3/TLS write certificate request
[server] SSL where: 8193 ret: 1 state: SSLv3/TLS write certificate
[server] SSL where: 8193 ret: 1 state: TLSv1.3 write server certificate verify
[server] SSL where: 8193 ret: 1 state: SSLv3/TLS write finished
[server] SSL where: 8193 ret: 1 state: TLSv1.3 early data
[server] SSL where: 8194 ret: -1 state: TLSv1.3 early data
[214419aac41463c6defd4348e9652eec] PerformIO: Result: -1 sslError: 2 sysError: 0 pendingOutputBefore: 0 pendingOutputAfter: 1078
[Server] [214419aac41463c6defd4348e9652eec] Performing SSL handshake: want_output
[Server] [214419aac41463c6defd4348e9652eec] Draining WriteBIO to socket
[Client] [b246df83a18436c2fd3ee572cf870586] Read 1078 bytes from socket
[Server] [214419aac41463c6defd4348e9652eec] Drained WriteBIO to socket
[Client] [b246df83a18436c2fd3ee572cf870586] Draining ReadBIO from unconsumedInput after socket read
[server] SSL where: 8194 ret: -1 state: TLSv1.3 early data
[214419aac41463c6defd4348e9652eec] PerformIO: Result: -1 sslError: 2 sysError: 0 pendingOutputBefore: 0 pendingOutputAfter: 0
[client] SSL where: 4097 ret: 1 state: SSLv3/TLS write client hello
[Server] [214419aac41463c6defd4348e9652eec] Performing SSL handshake: want_input_and_retry
[Server] [214419aac41463c6defd4348e9652eec] Waiting for socket to be readable
[client] SSL where: 4097 ret: 1 state: SSLv3/TLS read server hello
[client] SSL where: 4097 ret: 1 state: TLSv1.3 read encrypted extensions
[client] SSL where: 4097 ret: 1 state: SSLv3/TLS read server certificate request
[client] SSL where: 4097 ret: 1 state: SSLv3/TLS read server certificate
[client] SSL where: 4097 ret: 1 state: TLSv1.3 read server certificate verify
[client] SSL where: 4097 ret: 1 state: SSLv3/TLS read finished
[client] SSL where: 4097 ret: 1 state: SSLv3/TLS write change cipher spec
[client] SSL where: 4097 ret: 1 state: SSLv3/TLS write client certificate
[client] SSL where: 4097 ret: 1 state: SSLv3/TLS write finished
[client] SSL where: 32 ret: 1 state: SSL negotiation finished successfully
[client] SSL where: 4098 ret: 1 state: SSL negotiation finished successfully
[b246df83a18436c2fd3ee572cf870586] PerformIO: Result: 1 sslError: 0 sysError: 0 pendingOutputBefore: 0 pendingOutputAfter: 110
[Client] [b246df83a18436c2fd3ee572cf870586] Performing SSL handshake: want_output
[Client] [b246df83a18436c2fd3ee572cf870586] Draining WriteBIO to socket
[Client] [b246df83a18436c2fd3ee572cf870586] Drained WriteBIO to socket
[b246df83a18436c2fd3ee572cf870586] PerformIO: Result: 1 sslError: 0 sysError: 0 pendingOutputBefore: 0 pendingOutputAfter: 0
[Client] [b246df83a18436c2fd3ee572cf870586] Sending handshake successful
[Server] [214419aac41463c6defd4348e9652eec] Read 110 bytes from socket
[Server] [214419aac41463c6defd4348e9652eec] Draining ReadBIO from unconsumedInput after socket read
[server] SSL where: 8193 ret: 1 state: TLSv1.3 early data
[server] SSL where: 8193 ret: 1 state: SSLv3/TLS read client certificate
[Client] [b246df83a18436c2fd3ee572cf870586] FINAL HANDSHAKE SUCCEEDED
[server] SSL where: 8193 ret: 1 state: SSLv3/TLS read finished
[server] SSL where: 32 ret: 1 state: SSLv3/TLS write session ticket
[b246df83a18436c2fd3ee572cf870586] OnRead
[b246df83a18436c2fd3ee572cf870586] OnRead sslPointer=0x20e5f30
[b246df83a18436c2fd3ee572cf870586] SSLRead sslPointer: 0x20e5f30
[b246df83a18436c2fd3ee572cf870586] PerformIO: Result: -1 sslError: 2 sysError: 0 pendingOutputBefore: 0 pendingOutputAfter: 0
[b246df83a18436c2fd3ee572cf870586] Want input and retry
[b246df83a18436c2fd3ee572cf870586] Error while reading from socket: 0 Success
[b246df83a18436c2fd3ee572cf870586] SSLRead sslPointer: 0x20e5f30
[client] SSL where: 16392 ret: 582 state: error
[b246df83a18436c2fd3ee572cf870586] PerformIO: Result: -1 sslError: 1 sysError: 167772427 pendingOutputBefore: 0 pendingOutputAfter: 24
[b246df83a18436c2fd3ee572cf870586] Want output
[b246df83a18436c2fd3ee572cf870586] OnRead returned size: 0 error: 167772427 wrong version number (SSL routines)
[b246df83a18436c2fd3ee572cf870586] onReadError: 167772427 wrong version number (SSL routines)
[server] SSL where: 8193 ret: 1 state: SSLv3/TLS write session ticket
[b246df83a18436c2fd3ee572cf870586] Closing socket
[b246df83a18436c2fd3ee572cf870586] Socket closed
[server] SSL where: 8193 ret: 1 state: SSLv3/TLS write session ticket
[server] SSL where: 8194 ret: 1 state: SSL negotiation finished successfully
[214419aac41463c6defd4348e9652eec] PerformIO: Result: 1 sslError: 0 sysError: 0 pendingOutputBefore: 0 pendingOutputAfter: 510
[Server] [214419aac41463c6defd4348e9652eec] Performing SSL handshake: want_output
[Server] [214419aac41463c6defd4348e9652eec] Draining WriteBIO to socket
[Server] [214419aac41463c6defd4348e9652eec] Drained WriteBIO to socket
[214419aac41463c6defd4348e9652eec] PerformIO: Result: 1 sslError: 0 sysError: 0 pendingOutputBefore: 0 pendingOutputAfter: 0
[Server] [214419aac41463c6defd4348e9652eec] Sending handshake successful
[b246df83a18436c2fd3ee572cf870586] Closing socket
[b246df83a18436c2fd3ee572cf870586] Socket closed
[Server] [214419aac41463c6defd4348e9652eec] FINAL HANDSHAKE SUCCEEDED
[214419aac41463c6defd4348e9652eec] OnRead
[214419aac41463c6defd4348e9652eec] OnRead sslPointer=0x20e7170
[214419aac41463c6defd4348e9652eec] SSLRead sslPointer: 0x20e7170
[214419aac41463c6defd4348e9652eec] PerformIO: Result: -1 sslError: 2 sysError: 0 pendingOutputBefore: 0 pendingOutputAfter: 0
[214419aac41463c6defd4348e9652eec] Want input and retry
[214419aac41463c6defd4348e9652eec] Error while reading from socket: 0 Success
[214419aac41463c6defd4348e9652eec] SSLRead sslPointer: 0x20e7170
[server] SSL where: 16388 ret: 582 state: SSL negotiation finished successfully
[214419aac41463c6defd4348e9652eec] PerformIO: Result: 0 sslError: 1 sysError: 167773230 pendingOutputBefore: 0 pendingOutputAfter: 0
[214419aac41463c6defd4348e9652eec] OnRead returned size: 0 error: 167773230 tlsv1 alert protocol version (SSL routines)
[214419aac41463c6defd4348e9652eec] onReadError: 167773230 tlsv1 alert protocol version (SSL routines)
[214419aac41463c6defd4348e9652eec] Closing socket
[214419aac41463c6defd4348e9652eec] Socket closed
[214419aac41463c6defd4348e9652eec] Closing socket
[214419aac41463c6defd4348e9652eec] Socket closed

**Comment by nhorman** [comment, marked as answer]
do you have a tcpdump of this interaction you can share?  Ideally with an associated sslkeylog file to decode it?

**Reply by dreamvrutik** [reply]
> do you have a tcpdump of this interaction you can share?  Ideally with an associated sslkeylog file to decode it?

I have shared everything I could. This is a test being run on a docker container where I cannot access other things. Same test passes when instead of using OpenSSL we use boost asio
